### PR TITLE
[WIP] Two Factor Authentication

### PR DESCRIPTION
**Pull request by henninghaakonsen**

### CODE REVIEW

File: dhis-2/dhis-api/pom.xml [hunk 1]
```diff
@@ -107,6 +107,20 @@
...
+    <dependency>
+      <groupId>org.jboss.aerogear</groupId>
+      <artifactId>aerogear-otp-java</artifactId>
+      <version>1.0.0.CR1</version>
+    </dependency>
+    <dependency>
+      <groupId>org.springframework.security</groupId>
+      <artifactId>spring-security-core</artifactId>
+    </dependency>
+      <dependency>
```
Comment (larshelge) [inline dhis-2/dhis-api/pom.xml]: Codestyle, spaces. Move version to root pom dependencyManagement version.

File: dhis-2/dhis-services/dhis-service-core/pom.xml [hunk 2]
```diff
@@ -128,6 +128,23 @@
...
+    <!-- Security -->
+    <dependency>
+      <groupId>org.jboss.aerogear</groupId>
+      <artifactId>aerogear-otp-java</artifactId>
+      <version>1.0.0</version>
+    </dependency>
+    <dependency>
+      <groupId>javax.servlet</groupId>
+      <artifactId>javax.servlet-api</artifactId>
+      <scope>compile</scope>
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/pom.xml]: We don't need servlet API depedency here I think.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/SecurityUtils.java [hunk 3]
```diff
@@ -0,0 +1,59 @@
+package org.hisp.dhis.security;
+
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/SecurityUtils.java]: Add license

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/spring2fa/TwoFactorAuthenticationProvider.java [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+                log.info( String.format( "Two-factor authentication failure for user: %s", userCredentials.getUsername() ) );
+
+                throw new BadCredentialsException( "Invalid verification code" );
+            }
+        }
+
+        // -------------------------------------------------------------------------
+        // Delegate authentication downstream, using UserCredentials as principal
+        // -------------------------------------------------------------------------
+        Authentication result = super.authenticate( auth );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/spring2fa/TwoFactorAuthenticationProvider.java]: Space

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/spring2fa/TwoFactorWebAuthenticationDetails.java [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+    private static final String HEADER_FORWARDED_FOR = "X-Forwarded-For";
+
+    private String code;
+
+    private String ip;
+
+    TwoFactorWebAuthenticationDetails( HttpServletRequest request )
+    {
+        super( request );
+        code = request.getParameter( "2fa_code" );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/security/spring2fa/TwoFactorWebAuthenticationDetails.java]: Externalize string "2fa_code" to static string.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TwoFAPopulator.java [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+    public void execute()
+        throws Exception
+    {
+        UserQueryParams userQueryParams = new UserQueryParams( currentUserService.getCurrentUser() );
+        userQueryParams.setNot2FA( true );
+
+        userService.getUsers( userQueryParams ).forEach( user -> {
+            user.getUserCredentials().setSecret( null );
+            userService.updateUser( user );
+        } );
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/startup/TwoFAPopulator.java]: Nice solution

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/hibernate/HibernateUserStore.java [hunk 7]
```diff
@@ -99,6 +99,11 @@ private Query getUserQuery( UserQueryParams params, boolean count )
             hql += hlp.whereAnd() + " uc.disabled = :disabled ";
         }
 
+        if ( params.isNot2FA() )
+        {
+            hql += hlp.whereAnd() + " uc.secret is null ";
+        }
+
```
Comment (larshelge) [inline dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/hibernate/HibernateUserStore.java]: Should we also add a test for "twoFA" field here?

File: dhis-2/dhis-web/dhis-web-commons-resources/src/main/webapp/dhis-web-commons/security/login.js [hunk 8]
```diff
@@ -51,6 +61,7 @@ login.changeLocale = function( locale )
 		$( '#signInLabel' ).html( json.sign_in );
 		$( '#j_username' ).attr( 'placeholder', json.login_username );
 		$( '#j_password' ).attr( 'placeholder', json.login_password );
+        $( '#2fa_code' ).attr( 'placeholder', json.login_code );
```
Comment (larshelge) [inline dhis-2/dhis-web/dhis-web-commons-resources/src/main/webapp/dhis-web-commons/security/login.js]: Codestyle, indentation.

File: dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/hibernate/HibernateUserStore.java (hunk 7 above)
Comment (henninghaakonsen) [reply dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/user/hibernate/HibernateUserStore.java]: Not sure what purpose that would serve?

**Pull request merged**